cpp/src/toppra/algorithm.hpp

```diff
@@ -1,6 +1,106 @@
-#ifndef ALGORITHM_HPP
-#define ALGORITHM_HPP
+#ifndef TOPPRA_ALGORITHM_HPP
+#define TOPPRA_ALGORITHM_HPP
 
+#include <stdexcept>
+#include <toppra/constraint.hpp>
+#include <toppra/geometric_path.hpp>
+#include <toppra/solver.hpp>
+#include <toppra/toppra.hpp>
 
+namespace toppra {
+
+enum class ReturnCode {
+  OK = 0,
+  ERR_UNKNOWN = 1,
+  ERR_FAIL_CONTROLLABLE = 2,
+  ERR_FAIL_FORWARD_PASS = 3,
+  ERR_UNINITIALIZED = 4,
+};
+
+struct ParametrizationData {
+  /// \brief Grid-points used for solving the discretized problem.
+  /// The number of points must equal m_N + 1.
+  Vector gridpoints;
+
+  ///  Output parametrization (squared path velocity)
+  Vector parametrization;
+
+  Matrix controllable_sets;
+  Matrix feasible_sets;
+
+  ///  Return code of the algorithm.
+  ReturnCode ret_code = ReturnCode::ERR_UNINITIALIZED;
+};
+
+/** Base class for time parametrization algorithms.
+ *
+ */
+class PathParametrizationAlgorithm {
+ public:
+  /** Construct the problem instance.
+   *
+   *  \param  constraints  List of constraints.
+   *  \param  path  The geometric path.
+   *
+   */
+  PathParametrizationAlgorithm(LinearConstraintPtrs constraints,
+                               const GeometricPath &path);
+
+  /** \brief Set the level of discretization used by the solver.
+   *
+   * If is zero, will attempt to detect automatically the most suitable grid.
+   */
+  void setN(int N) { m_N = N; };
+
+  /** \brief Get output or result of algorithm.
+   */
+  ParametrizationData getParameterizationData() const { return m_data; };
+
+  /** Compute the time parametrization of the given path.
+   *
+   * \param vel_start
+   * \param vel_end
+   * \return Return code.
+   */
+  virtual ReturnCode computePathParametrization(double vel_start = 0,
+                                                double vel_end = 0);
+  virtual ~PathParametrizationAlgorithm() {}
+
+ protected:
+  /** \brief Select solver and gridpoints to use.
+   *
+   * This method implements a simple way to select gridpoints.
+   */
+  virtual void initialize();
+
+  /** \brief Compute the forward pass.
+   *
+   * Derived class should provide a suitable forward pass function,
+   * depending on the desired objective.
+   */
+  virtual ReturnCode computeForwardPass(double vel_start) = 0;
+
+  /** Compute the sets of feasible squared velocities.
+   */
+  ReturnCode computeFeasibleSets() { throw std::runtime_error("Not implemented"); };
```

Comment:
Is it possible to write a PathParameterizationAlgorithm with implementing this ? If no, make it pure virtual, otherwise forget about this comment.

Reply:
It's possible. Will do.